### PR TITLE
[#12457] Make the table buttons easier to use on mobile devices

### LINKED ISSUE #12457: Make the table buttons easier to use on mobile devices
Issue title: Change the instructor student buttons change on mobile devices

- **Environment**: The environment in which the said feature/enhancement is absent is the master branch

**Description of feature/enhancement**

Currently when an instructor is at the students page the actions buttons have different sizes when the user is logged in from a mobile device. This is not very aesthetically pleasant for the user. The UI is given below:

new issue before

After some time and many alterations to the code, I managed to change the actions buttons to be all the same size **only when the user is logged in from a mobile screen**. The change of the UI can be seen below:

new issue after

**Justification**
The instructors will benefit from this change as the page seems more "clean" , organized, and professional. I added a css class where the size of the button (width and height) can change so that the button can become smaller or larger if it is required by the TEAMMATES team, and I also added the necessary code so that the respective buttons refer to the respective css class

### PR DESCRIPTION
**Pull request by athakaras**
Fixes #12457 

**Outline of Solution**

I fixed the buttons so that all buttons have the same sizing and have some spacing in between them so that the UI is more clean and readable by the user. These changes were implemented through a scss class that included the necessary code and is referenced in each button that the changes must be implemented.

The pages include *instructor-courses page* **for both mobile and desktop screens** as well as the *instructor-student page and the student-home page* **only for mobile screens**.

All implemented changes in the UI are shown in the following screenshots:
students-home
instructor-students

### CODE REVIEW

**Comment by weiquu** [comment]
Hi @athakaras, some preliminary comments for each page below. Do voice out any disagreements or suggestions! I think this issue allows for a fairly wide range of potential solutions.

**Instructor courses page**
In some screen sizes, the buttons get pushed to different rows despite having enough space to fit into a single row. This leaves quite a bit of space by the sides (see below, at 1123px width). Perhaps we could have the button size be set to fill the row, subject to a minimum size? I'm not too sure about this, other opinions would be very welcome.

Original:
Screenshot 2023-06-07 at 5 09 05 PM
Changed:
Screenshot 2023-06-07 at 5 05 54 PM

Also noted that there appears to be some slight misalignment between the upper and lower buttons.


**Instructor student page**
Let's add a margin between lines whenever it appears (below example at 777px width):
Screenshot 2023-06-07 at 5 06 47 PM

Note that the above suggestion of setting the button size to fill the row (with a min size enforced) would enable standardisation of button sizes here (e.g. Delete button).


**Student home page**
Not sure why my page looks like this (desktop view):
Screenshot 2023-06-07 at 5 06 04 PM

Could you help to take a look?

**Comment by weiquu** [comment]
Hi @athakaras, thanks for looking into the issues!

**Instructor courses page**
Unfortunately I won't be able to take a look at this until Friday/the weekend, so I'm not too sure what's causing the misalignment. My guess is that it's due to some extra whitespace, or some added left/right margin to the button. If you can't figure it out then no worries for now!

**Instructor student page**
Thanks for pointing out! It seems that this lack of margin between rows issue is present in quite a few other areas. Since it's not directly related to this PR, we can leave it to another time

**Student home page**
You'll have to go to the instructor pages and either extend the session time to a later date (i.e. later than today), or create a new session. If you're creating a new session, you'll have to go back to the student pages and submit your answers for that session, after which the edit submission button should appear. Do let me know if this doesn't work!

**Bootstrap and breakpoints**
Noted! I think the styles added in are fine, so let's keep it that way for now. For the breakpoint you put it as max width 767px, but I think most styles currently use a max width of 768px... doesn't make too much of a difference, but maybe we should standardise to 768px? @domlimm lmk what you think of this and the rest of the stuff above (:

**Comment by domlimm** [comment]
@athakaras @weiquu so sorry for getting back late. I will just add my comments down below.

> For the breakpoint you put it as max width 767px, but I think most styles currently use a max width of 768px... doesn't make too much of a difference, but maybe we should standardise to 768px?

Yes, let's keep it to 768px please! This from [w3schools](https://www.w3schools.com/css/css_rwd_mediaqueries.asp) might be helpful.

> Some thoughts

Based on all that were discussed, in my opinion, the buttons should try to fill up the spaces equally and symmetrically i.e., All alignment are centered, and some spacing at the left and right WHEN the screen size gets smaller. I think it might be difficult to understand what I just said but it's exactly what @weiquu had commented above in the 1st comment. Let's try to do that wherever possible 👍🏻 

> Let's try to use Bootstrap [classes](https://getbootstrap.com/docs/5.0/components/buttons/) and [breakpoints](https://getbootstrap.com/docs/5.0/layout/breakpoints/) as far as possible! Do quickly check if there's anything there that can be changed.

Always the best guideline as we're using external libraries to help us with the UI.

I believe that should be it. Thank you both! @athakaras do ping us once you're ready for a review, thanks!

**Comment by athakaras** [comment]
Hello @domlimm @weiquu ! Thanks for the help. These are what I found based on your feedback:

**Instructor courses page**

I think I found what causes the misalignment. Both buttons are referenced to the same scss class and have the same size, however the *other actions* button also has an arrow shape that makes the button a bit widther and that is what causes the misalignment. I am not sure how to fix this.

**Bootstrap and breakpoints**

I changed the media px from 767 to 768 according to the standardised styles of css everywhere so I think now the fix is okay

**Student home page**

I added the scss class that I created also to the *edit submission* and *start submission* buttons so now the fix is okay and all the buttons have the same sizing.

**Overall**

I think I included everything that you told me so please proceed with your review and let me know if I have to change anything else before merging. Thank you for your time!!

**Review by weiquu — CHANGES_REQUESTED** [review]
Hi @athakaras, looks good so far, but do take a look at the failing tests and then we'll take a more detailed look at the code (although I don't foresee any issues).

Couple more comments:

The misalignment issue in the **instructor courses page** seems to be caused by lines 87-89 in the `instructor-courses-page.component.scss`:
```
.actions-cell button:not(.dropdown-item) {
  margin-left: var(--btn-margin);
}
```
A possible solution is to just apply it to non-mobile screens. A quick search shows that most `.scss` files use `min-width: 768px` along with `max-width: 768px`... which I believe runs into the odd issue when the screen width is exactly 768px. Haven't really investigated this, but if it is indeed a problem, we can leave it to another issue/PR. In summary, feel free to use `min-width: 768px` for the above for now (:

On the same page, the buttons for the Archived courses and Deleted courses tables also face a similar issue. It would be good to fix those as well.

Last thing: shall we centre-align the buttons in the **student home page**? I think they weren't centre-aligned previously... not too sure if there was a reason for that.

**Comment by weiquu** [comment]
Thanks @athakaras, could you provide a screenshot of the issue with the Delete buttons? But in any case, I think we can leave that to a future PR; let's try to set the groundwork first with this one.

For the student home button misalignment, seems that the issue is caused by the `.btn-margin-right` class, but we can't remove that also, else the buttons will be too close together when they're on a single line. Can't really think of any solutions now since the buttons don't seem to revert to a single line at any particular breakpoint... @domlimm any suggestions? If not, we can also raise it as a separate issue since I'm sure it'll come up when fixing other buttons for mobile screens.

Anyway, I believe nothing else needed on your end for now! Do give us a few days to do the review proper (:

File: src/web/app/components/student-list/student-list.component.scss
Comment (weiquu) [inline src/web/app/components/student-list/student-list.component.scss:7]: Do you think we could remove this `font-size`? I toggled it on and off and couldn't tell the difference; prefer keeping to the `0.875rem` defined since there's some [accessibility-related concerns](https://austingil.com/px-or-rem-in-css/).

File: src/web/app/components/student-list/student-list.component.scss
Comment (weiquu) [inline src/web/app/components/student-list/student-list.component.scss:9]: Don't think the height of the button was an issue before, and without this line the button height seems to increase... shall we remove it as well?

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.scss
Comment (weiquu) [inline src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.scss:18]: Placing a few comments together:
- If the above comment on the height of the button makes sense, let's remove both the `height`s here as well
- Do delete the commented line if it's not needed!
- Let's rename `.custom-buttons` as `.custom-button-active` and `.custom-buttons2` as `.custom-button-archived`. I think that'll be a bit more descriptive (don't forget to make the corresponding changes in the `.html` file)

File: src/web/app/pages-student/student-home-page/student-home-page.component.scss
Comment (weiquu) [inline src/web/app/pages-student/student-home-page/student-home-page.component.scss:9]: Was thinking if we could follow the same format as the buttons in the instructor pages:
- Set the styles to apply to a `max-width` of `768px`
- Remove the `height` if not necessary
- Change the name to `.custom-button` or something more descriptive if possible (with corresponding change in the `.html` file)

The above change also allows us to remove the misalignment by adding `margin-right: 0;` for mobile screens

**Review by weiquu — CHANGES_REQUESTED** [review]
Hi @athakaras, couple of minor changes! Just some renaming and checking if all the styles applied are needed. Probably the biggest change is to the student page - more details in the comments (:

**Comment by domlimm** [comment]
@athakaras @weiquu hey guys, sorry for the late reply. Will get to the ping tonight.

**Comment by athakaras** [comment]
@weiquu @domlimm I changed everything that @domlimm proposed so now the code is better written. Also in the *student-home-page* I added a margin right as proposed so the misalignment is now fixed!

I hope everything is okay now!!

**Review by weiquu — APPROVED** [review]
LGTM! Thanks for all the changes (: Pinging @domlimm

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.scss
Comment (domlimm) [inline src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.scss:95]: @athakaras @weiquu Not really proficient in advanced CSS, may I ask where is `--btn-margin` coming from? Or rather what is the value of this variable?

**Review by domlimm — APPROVED** [review]
@athakaras @weiquu So sorry for the late review. Just took some time to look at everything.

I noticed that when width is > 768px, the buttons (at least in instructor courses page) become awkward looking again. It's fine >= 1079px onwards. To sum up, 769-1078px does need some tweaking. Maybe we can leave it for another issue.

Another thing - is this meant to look like that? The 2 buttons do not have any spacing between them @ 767px, and when courses are archived.

image

All in all, LGTM! Thank you for making all the changes, well done @athakaras and @weiquu for helping all the way! Kudos to you both! 😃

Will wait on your replies before merging.

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.scss
Comment (weiquu) [reply src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.scss:95]: @domlimm I believe it's from `styles.scss` where it was set with:
```
:root {
  --btn-margin: 5px;
}
```

Can read more about it [here](https://developer.mozilla.org/en-US/docs/Web/CSS/var)!

**Comment by weiquu** [comment]
> I noticed that when width is > 768px, the buttons (at least in instructor courses page) become awkward looking again. It's fine >= 1079px onwards. To sum up, 769-1078px does need some tweaking. Maybe we can leave it for another issue.

I think let's leave the various spacing problems to another issue... we'll keep this PR as more of a groundwork towards making buttons easier to use on small screens.

**Pull request merged**
